### PR TITLE
storage: turn off panic on consistency failure

### PR DESCRIPTION
**Pull request by vivekmenezes**
Added COCKROACH_CONSISTENCY_CHECK_PANIC_ON_FAILURE to turn on
panic in acceptance tests and beta clusters.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5687)

### CODE REVIEW

**Comment by vivekmenezes** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/replica_command.go, line 1620 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE3-hNBraeLmWZA7DGH:-KE3-hNCRXzTAzm2H1Qx:1857817810) ([raw file](https://github.com/cockroachdb/cockroach/blob/9157a9e4f1800601267c728c39ae742c123835bc/storage/replica_command.go#L1620)):*
It would be nicer to read this into the server context but unfortunately a replica has no access to the server context. One way to do that is to give the store access to the server context.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687)*

**Comment by petermattis** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/replica_command.go, line 1626 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE30p4ladZT0YQsscK7:-KE30p4ladZT0YQsscK8:736833132) ([raw file](https://github.com/cockroachdb/cockroach/blob/9157a9e4f1800601267c728c39ae742c123835bc/storage/replica_command.go#L1626)):*
`strconv.ParseBool` returns false for the first argument if it can't be parsed. Instead of dying here, I think you should be permissive and do something like:

```
panicOnConsistencyFailure := strconv.ParseBool(os.Getenv("COCKROACH_PANIC_ON_CONSISTENCY_FAILURE"))
```

---

*[storage/replica_command.go, line 1633 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE30Q2avUK8_QJO1IQ1:-KE30Q2bgz6raJr_iHAa:966563705) ([raw file](https://github.com/cockroachdb/cockroach/blob/9157a9e4f1800601267c728c39ae742c123835bc/storage/replica_command.go#L1633)):*
This is fine, though you could invert the logic and do something like:

```
logFunc := log.Error
if shouldPanic {
  logFunc = log.Fatal
}
logFunc("replica: %s, checksum mismatch ...", ...)
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-:-KE317nAEYmAZ5mi9qhg:1417088721)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/replica_command.go, line 1626 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE30p4ladZT0YQsscK7:-KE36eZHweJTrX0BE8I5:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/9157a9e4f1800601267c728c39ae742c123835bc/storage/replica_command.go#L1626)):*
Done

---

*[storage/replica_command.go, line 1633 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE30Q2avUK8_QJO1IQ1:-KE36fjPIz-BrXBxzc3D:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/9157a9e4f1800601267c728c39ae742c123835bc/storage/replica_command.go#L1633)):*
Done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687)*

File: storage/replica_command.go [hunk 1]
```diff
@@ -1595,7 +1597,14 @@ func (r *Replica) VerifyChecksum(
 			p()
 		} else {
 			// Replication consistency problem!
-			log.Fatalf("replica: %s, checksum mismatch: e = %x, v = %x", r, args.Checksum, c.checksum)
+
+			// shouldPanic is set to false when the env var is unset, or fails parsing.
+			shouldPanic, _ := strconv.ParseBool(os.Getenv("COCKROACH_PANIC_ON_CONSISTENCY_FAILURE"))
```
Comment (BramGruneir) [inline storage/replica_command.go]: Shouldn't this be load in the server context?
https://github.com/cockroachdb/cockroach/blob/master/server/context.go
See readEnvironmentVariables()

Comment (vivekmenezes) [inline storage/replica_command.go]: See my comment about the replica not having access to the server context.

**Comment by andreimatei** [comment]
Shouldn't this be on by default?

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/replica_command.go, line 1620 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE3-hNBraeLmWZA7DGH:-KE38_034icosG7k01a0:-366197145) ([raw file](https://github.com/cockroachdb/cockroach/blob/9157a9e4f1800601267c728c39ae742c123835bc/storage/replica_command.go#L1620)):*
Isn't there a store context created from the server context? I think this is how we plumb other setting around.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-:-KE38fSAKpdarrGZhvlT:125925438)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/replica_command.go, line 1620 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE3-hNBraeLmWZA7DGH:-KE39INvzX3MXgMWZ33A:-2087511625) ([raw file](https://github.com/cockroachdb/cockroach/blob/9157a9e4f1800601267c728c39ae742c123835bc/storage/replica_command.go#L1620)):*
+1 this needs to be plumbed correctly and defined in the same place where we defined other environment variables.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687)*

File: storage/replica_command.go (hunk 1 above)
Comment (BramGruneir) [inline storage/replica_command.go]: I didn't see this comment.  I can't see it in the file.  

But a replica has a store, the store has a context, should be easy to add it in there.  ScanMaxIdleTime is in there which is plumbed from server.context.

**Comment by vivekmenezes** [comment]
It's off by default because we're still hitting issues.

---

Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/replica_command.go, line 1620 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE3-hNBraeLmWZA7DGH:-KE433EnmGmG-Vo4xRJW:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/9157a9e4f1800601267c728c39ae742c123835bc/storage/replica_command.go#L1620)):*
Done

---

*[storage/replica_command.go, line 1602 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE3-ciDRUy0b5WKw7SI-r2-1602:-KE435By7cQ390w1OdBU:-497468049) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca30e33a49932077cb7c231ce4c030c121b81b13/storage/replica_command.go#L1602)):*
Fixed

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-:-KE438-xzR_wlrRwqSw3:482509692)*

File: server/context.go [hunk 2]
```diff
@@ -318,22 +322,35 @@ func parseDurationEnv(env, internalName string, duration *time.Duration) {
 // when NewContext is called.
 func (ctx *Context) readEnvironmentVariables() {
 	// cockroach-linearizable
-	if linearizableString := os.Getenv("COCKROACH_LINEARIZABLE"); len(linearizableString) != 0 {
-		if linearizable, err := strconv.ParseBool(linearizableString); err != nil {
-			log.Errorf("could not parse environment variable COCKROACH_LINEARIZABLE=%s, setting to default of %t, error: %s",
-				linearizableString, ctx.Linearizable, err)
+	linearizableEnv := "COCKROACH_LINEARIZABLE"
+	if value := os.Getenv(linearizableEnv); len(value) != 0 {
```
Comment (BramGruneir) [inline server/context.go]: How about combining these two boolean checks into a parseBoolEnv function to be more DRY?

**Comment by BramGruneir** [comment]
LGTM with that one last change.

**Comment by vivekmenezes** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[server/context.go, line 326 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE42xaH9cjEz9lUtA2U-r3-326:-KE6q2zEQe4S08ii6H4N:1367071830) ([raw file](https://github.com/cockroachdb/cockroach/blob/469a28bb182b7ac6a9a6483c0f3134f303c7d126/server/context.go#L326)):*
Good suggestion. Done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687)*

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 4 of 6 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[server/context.go, line 150 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE6rv--k35EZXNClurw:-KE6rv--k35EZXNClurx:-1255129695) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8d2398107544fbc5da069f8f855987bc736ff44/server/context.go#L150)):*
s/allows/causes/

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-:-KE6t29TsmXVU8OA59JY:1417088721)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 362 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE6t4GOFPRkbGkOmlEv:-KE6t4GOFPRkbGkOmlEw:-449644553) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b4d367217f8e8b3033dc487dea0ea5d07f2aabd/storage/store.go#L362)):*
ditto s/allows/causes/

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687)*

**Comment by vivekmenezes** [comment]
Review status: 4 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[server/context.go, line 150 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE6rv--k35EZXNClurw:-KE6y78MMXJYoh7qCcsk:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8d2398107544fbc5da069f8f855987bc736ff44/server/context.go#L150)):*
done

---

*[storage/store.go, line 362 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687#-KE6t4GOFPRkbGkOmlEv:-KE6y5NKWtE9vBjzvZjv:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b4d367217f8e8b3033dc487dea0ea5d07f2aabd/storage/store.go#L362)):*
done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5687)*

**Pull request merged**